staging/src/k8s.io/apiserver/pkg/storage/testing/store_tests.go

```diff
@@ -619,6 +621,248 @@ func RunTestPreconditionalDeleteWithOnlySuggestionPass(ctx context.Context, t *t
+			})
+		}
+		b.Run("Paginate", func(b *testing.B) {
+			for _, c := range cases {
+				b.Run(c.name, func(b *testing.B) {
+					runBenchmarkStoreList(ctx, b, store, paginateLimit, maxRevision, c.match, true, nodeNames)
+				})
+			}
+		})
+	})
+	b.Run("Namespace", func(b *testing.B) {
+		for _, c := range cases {
+			b.Run(c.name, func(b *testing.B) {
+				runBenchmarkStoreListNamespace(ctx, b, store, maxRevision, c.match, namespacedNames)
+			})
+		}
+	})
+}
+
+func runBenchmarkStoreListNamespace(ctx context.Context, b *testing.B, store storage.Interface, maxRV int, match metav1.ResourceVersionMatch, namespaceNames []string) {
+	wg := sync.WaitGroup{}
+	objectCount := atomic.Uint64{}
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		wg.Add(1)
+		resourceVersion := ""
+		switch match {
+		case metav1.ResourceVersionMatchExact:
+			resourceVersion = fmt.Sprintf("%d", maxRV-99+i%100)
+		case metav1.ResourceVersionMatchNotOlderThan:
+			resourceVersion = "0"
+		}
+		go func(resourceVersion string) {
+			defer wg.Done()
+			for j := 0; j < len(namespaceNames); j++ {
+				listOut := &example.PodList{}
+				err := store.GetList(ctx, "/pods/"+namespaceNames[j], storage.ListOptions{
+					Recursive:            true,
+					ResourceVersion:      resourceVersion,
+					ResourceVersionMatch: match,
+					Predicate:            storage.Everything,
+				}, listOut)
+				if err != nil {
+					panic(fmt.Sprintf("Unexpected error %s", err))
+				}
+				objectCount.Add(uint64(len(listOut.Items)))
+			}
+		}(resourceVersion)
+	}
+	wg.Wait()
+	b.ReportMetric(float64(objectCount.Load())/float64(b.N), "objects/op")
+}
+
+func runBenchmarkStoreList(ctx context.Context, b *testing.B, store storage.Interface, limit int64, maxRV int, match metav1.ResourceVersionMatch, perNode bool, nodeNames []string) {
+	wg := sync.WaitGroup{}
+	objectCount := atomic.Uint64{}
+	pageCount := atomic.Uint64{}
+	for i := 0; i < b.N; i++ {
+		wg.Add(1)
+		resourceVersion := ""
+		switch match {
+		case metav1.ResourceVersionMatchExact:
+			resourceVersion = fmt.Sprintf("%d", maxRV-99+i%100)
+		case metav1.ResourceVersionMatchNotOlderThan:
+			resourceVersion = "0"
+		}
+		go func(resourceVersion string) {
+			defer wg.Done()
+			if perNode {
+				for _, nodeName := range nodeNames {
+					field := fields.SelectorFromSet(fields.Set{"spec.nodeName": nodeName})
+					paginate(ctx, store, &objectCount, &pageCount, limit, resourceVersion, match, field)
+				}
+			} else {
+				paginate(ctx, store, &objectCount, &pageCount, limit, resourceVersion, match, fields.Everything())
+			}
+		}(resourceVersion)
+	}
+	wg.Wait()
+	b.ReportMetric(float64(objectCount.Load())/float64(b.N), "objects/op")
+	b.ReportMetric(float64(pageCount.Load())/float64(b.N), "pages/op")
+}
+
+func paginate(ctx context.Context, store storage.Interface, objectCount, pageCount *atomic.Uint64, limit int64, resourceVersion string, match metav1.ResourceVersionMatch, fields fields.Selector) {
```

Comment:
I'm not a fan of passing objectCount/pageCount as *atomic.Uint64,

It would be much cleaner to have:
`func paginate(ctx, store, limit, rv, match, fields) (uint64, uint64)`

Reply:
Or going even further:
```
func paginate(ctx, store, listOptions) (uint64, uint64)
```
then for the first call you just use the passed options and for subsequent you clear our RV and RVM and set Continue

Reply:
Done